e2e/v3/cluster3/cluster3.go

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package cluster3
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	consulapi "github.com/hashicorp/consul/api"
+	nomadapi "github.com/hashicorp/nomad/api"
+	vaultapi "github.com/hashicorp/vault/api"
+	"github.com/shoenig/test/must"
+	"github.com/shoenig/test/wait"
+)
+
+type Cluster struct {
+	t *testing.T
+
+	consulClient *consulapi.Client
+	nomadClient  *nomadapi.Client
+	vaultClient  *vaultapi.Client
+
+	timeout        time.Duration
+	leaderReady    bool
+	consulReady    bool
+	vaultReady     bool
+	linuxClients   int
+	windowsClients int
+}
+
+func (c *Cluster) wait() {
+	errCh := make(chan error)
+
+	statusAPI := c.nomadClient.Status()
+	nodesAPI := c.nomadClient.Nodes()
+	consulStatusAPI := c.consulClient.Status()
+	vaultSysAPI := c.vaultClient.Sys()
+
+	waitLeader := wait.InitialSuccess(
+		wait.Timeout(c.timeout),
+		wait.Gap(1*time.Second),
+		wait.TestFunc(func() (bool, error) {
+			if !c.leaderReady {
+				return true, nil
+			}
+			result, err := statusAPI.Leader()
+			return result != "", err
+		}),
+	)
+
+	waitLinuxClients := wait.InitialSuccess(
+		wait.Timeout(c.timeout),
+		wait.Gap(1*time.Second),
+		wait.ErrorFunc(func() error {
+			if c.linuxClients <= 0 {
+				return nil
+			}
+			nodes, _, err := nodesAPI.List(nil)
+			if err != nil {
+				return err
+			}
+			eligible := 0
+			for _, node := range nodes {
+				if node.Status == "ready" {
+					info, _, err := nodesAPI.Info(node.ID, nil)
+					if err != nil {
+						return err
+					}
+					if info.Attributes["kernel.name"] == "linux" {
+						eligible++
+					}
+				}
+			}
```

Comment:
Code golf: we could probably add a filter on the `List` API call for ready node with the appropriate attribute.